src/compiler/checker.ts

```diff
@@ -6743,15 +6775,28 @@ module ts {
                 if (arg.kind !== SyntaxKind.OmittedExpression) {
                     let paramType = getTypeAtPosition(signature, i);
                     let argType: Type;
-                    if (i === 0 && args[i].parent.kind === SyntaxKind.TaggedTemplateExpression) {
+                    if (arg.parent.kind === SyntaxKind.Decorator) {
+                        if (i === 0) {
+                            argType = getDecoratorTargetArgumentType(arg.parent.parent);
+                        }
+                        else if (i === 1) {
+                            argType = getDecoratorPropertyKeyArgumentType(arg.parent.parent);
+                        }
+                        else if (i === 2) {
+                            argType = getDecoratorPropertyIndexOrDescriptorArgumentType(arg.parent.parent);
+                        }
+                    }
+                    else if (i === 0 && arg.parent.kind === SyntaxKind.TaggedTemplateExpression) {
                         argType = globalTemplateStringsArrayType;
                     }
-                    else {
+
+                    if (argType === undefined) {

```

Comment:
How will this work? You are saying that even in the case of decorators or tagged template expressions, you are going to check the expression given by the effective call arguments. But I don't think that is correct. You will get `decorator.expression` for all of these, which is not the argument you are trying to infer from. I think the `else` behavior was correct.